File: apps/desktop/electron/shell-viewport.ts
import { createHash } from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'

import { type UguiDocument, validateUguiDocument } from '@hermes/shared/ugui-document'

const SAFE_ID_RE = /^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/

const PLATFORM_BY_SHELL: Record<string, string[]> = {
  'android-shell': ['android'],
  'linux-shell': ['linux'],
  'macos-shell': ['macos', 'ios'],
  'windows-shell': ['windows']
}

export interface ShellViewportSource {
  builds: Record<string, unknown>
  capabilities: Record<string, unknown>
  surfaces: Record<string, unknown>
}

export interface ShellViewportRequest {
  shell_id: string
  surface_profile_id: string
  target_id: string
}

export interface ShellViewportModel {
  schema: 'ae-shell-viewport-model/1'
  authority: 'none'
  source_hashes: {
    builds: `sha256:${string}`
    capabilities: `sha256:${string}`
    surfaces: `sha256:${string}`
  }
  shell: {
    id: string
    owner: string
    platform: string
    manifest: string
  }
  surface: {
    id: string
    name: string
    form_factor: 'desktop' | 'handset' | 'spatial' | 'wearable'
    production_hosts: string[]
    geometry: Record<string, unknown>
    safe_area: Record<string, unknown>
    corner_radii: Record<string, unknown>
    chrome: string[]
    window_policy: string
    sources: Array<Record<string, unknown>>
  }
  target: {
    id: string
    cell_id: string
    architecture: string
    sdk: string
    abi: null | string
    package: string
    artifact_kind: string
    disposition: string
    owner_ref: string
    reason: string
    rungs: {
      source: string
      artifact: string
      package_install: string
      physical_runtime: string
    }
  }
  capability_summary: Record<'available' | 'degraded' | 'unavailable' | 'unknown', number>
  capability_sample: Array<{
    id: string
    availability: 'available' | 'degraded' | 'unavailable' | 'unknown'
    owner: string
    reason: string
  }>
  posture: 'structural-projection'
  warning: string
  selector: {
    shells: string[]
    surfaces: string[]
    targets: string[]
  }
}

const object = (value: unknown): value is Record<string, unknown> => Boolean(value && typeof value === 'object' && !Array.isArray(value))
const array = (value: unknown): unknown[] => (Array.isArray(value) ? value : [])
const string = (value: unknown): string => (typeof value === 'string' ? value : '')

const hash = (value: unknown): `sha256:${string}` =>
  `sha256:${createHash('sha256').update(JSON.stringify(value)).digest('hex')}`

export function loadShellViewportSource(directory: string): ShellViewportSource {
  const read = (name: string): Record<string, unknown> => {
    const file = path.join(directory, name)
    const bytes = fs.readFileSync(file)

    if (bytes.length === 0 || bytes.length > 2 * 1024 * 1024) {throw new Error('shell-viewport-source-bound')}
    const value: unknown = JSON.parse(bytes.toString('utf8'))

    if (!object(value)) {throw new Error('shell-viewport-source-invalid')}

    return value
  }

  return {
    builds: read('shell-builds.json'),
    capabilities: read('shell-capability-parity.json'),
    surfaces: read('surface-profiles.json')
  }
}

function safeRequest(request: ShellViewportRequest): void {
  if (![request.shell_id, request.surface_profile_id, request.target_id].every(value => SAFE_ID_RE.test(value))) {
    throw new Error('shell-viewport-id')
  }
}

function platformForCapability(buildPlatform: string): string {
  if (buildPlatform === 'linux-desktop') {return 'linux'}

  if (buildPlatform === 'windows-desktop') {return 'windows'}

  if (['visionos', 'tvos', 'watchos'].includes(buildPlatform)) {return 'ios'}

  return buildPlatform
}

function compatible(shellId: string, buildPlatform: string, productionHosts: string[]): boolean {
  const capabilityPlatform = platformForCapability(buildPlatform)

  if (!(PLATFORM_BY_SHELL[shellId] ?? []).includes(capabilityPlatform)) {return false}

  if (shellId === 'android-shell') {return productionHosts.includes('android-shell')}

  if (shellId === 'macos-shell') {return productionHosts.some(host => host.startsWith('macos') || host.startsWith('ios') || host.startsWith('apple-'))}

  return true
}

export function buildShellViewportModel(source: ShellViewportSource, request: ShellViewportRequest): ShellViewportModel {
  safeRequest(request)
  const builds = source.builds
  const capabilities = source.capabilities
  const surfaces = source.surfaces

  if (
    builds.schema !== 'ae-shell-build-matrix/1' ||
    capabilities.schema !== 'ae-shell-capability-parity/1.0.0' ||
    surfaces.schema !== 'ugui-surface-profiles/1'
  ) {throw new Error('shell-viewport-source-schema')}

  const shell = array(builds.shells).find(row => object(row) && row.id === request.shell_id)
  const surface = array(surfaces.profiles).find(row => object(row) && row.id === request.surface_profile_id)

  const target = array(builds.platforms)
    .filter(object)
    .flatMap(platform => array(platform.targets).filter(object).map(row => ({ platform, row })))
    .find(entry => entry.row.id === request.target_id)

  const cell = array(builds.cells).find(row => object(row) && row.shell === request.shell_id && row.target === request.target_id)

  if (!object(shell) || !object(surface) || !target || !object(cell)) {throw new Error('shell-viewport-selection')}

  const buildPlatform = string(target.platform.id)
  const productionHosts = array(surface.productionHosts).map(string).filter(Boolean)

  if (!compatible(request.shell_id, buildPlatform, productionHosts)) {throw new Error('shell-surface-incompatible')}

  const capabilityPlatform = platformForCapability(buildPlatform)

  const capabilityRows = array(capabilities.capabilities)
    .filter(object)
    .map(row => ({ row, platform: object(row.platforms) && object(row.platforms[capabilityPlatform]) ? row.platforms[capabilityPlatform] as Record<string, unknown> : null }))
    .filter((entry): entry is { row: Record<string, unknown>; platform: Record<string, unknown> } => Boolean(entry.platform))

  const counts: ShellViewportModel['capability_summary'] = { available: 0, degraded: 0, unavailable: 0, unknown: 0 }

  for (const entry of capabilityRows) {
    const availability = string(entry.platform.availability) as keyof typeof counts

    if (!(availability in counts)) {throw new Error('shell-capability-availability')}
    counts[availability] += 1
  }

  const sample = capabilityRows
    .filter(entry => string(entry.platform.availability) !== 'unknown')
    .slice(0, 8)
    .map(entry => ({
      id: string(entry.row.capability),
      availability: string(entry.platform.availability) as ShellViewportModel['capability_sample'][number]['availability'],
      owner: string(entry.platform.owner),
      reason: string(entry.platform.reason)
    }))

  const qualification = object(cell.qualification_state) ? cell.qualification_state : {}

  const shellTargets = array(builds.cells)
    .filter(row => object(row) && row.shell === request.shell_id)
    .map(row => string((row as Record<string, unknown>).target))
    .filter(Boolean)

  const compatibleSurfaces = array(surfaces.profiles)
    .filter(object)
    .filter(row => compatible(request.shell_id, buildPlatform, array(row.productionHosts).map(string)))
    .map(row => string(row.id))

  const shellIds = array(builds.shells).filter(object).map(row => string(row.id))

  return {
    schema: 'ae-shell-viewport-model/1',
    authority: 'none',
    source_hashes: { builds: hash(builds), capabilities: hash(capabilities), surfaces: hash(surfaces) },
    shell: {
      id: request.shell_id,
      owner: string(shell.owner),
      platform: buildPlatform,
      manifest: string(shell.manifest)
    },
    surface: {
      id: request.surface_profile_id,
      name: string(surface.name),
      form_factor: string(surface.formFactor) as ShellViewportModel['surface']['form_factor'],
      production_hosts: productionHosts,
      geometry: object(surface.geometry) ? surface.geometry : {},
      safe_area: object(surface.safeArea) ? surface.safeArea : {},
      corner_radii: object(surface.cornerRadii) ? surface.cornerRadii : {},
      chrome: array(surface.chrome).map(string),
      window_policy: string(surface.windowPolicy),
      sources: array(surface.sources).filter(object)
    },
    target: {
      id: request.target_id,
      cell_id: string(cell.id),
      architecture: string(target.row.architecture),
      sdk: string(target.row.sdk),
      abi: typeof target.row.abi === 'string' ? target.row.abi : null,
      package: string(target.row.package),
      artifact_kind: string(target.row.artifact_kind),
      disposition: string(cell.disposition),
      owner_ref: string(cell.artifact_owner),
      reason: string(cell.reason),
      rungs: {
        source: string(qualification.source),
        artifact: string(qualification.artifact),
        package_install: string(qualification.package_install),
        physical_runtime: string(qualification.physical_runtime)
      }
    },
    capability_summary: counts,
    capability_sample: sample,
    posture: 'structural-projection',
    warning: 'STRUCTURAL PROJECTION — NOT A PHYSICAL RUN',
    selector: { shells: shellIds, surfaces: compatibleSurfaces, targets: shellTargets }
  }
}

const action = (id: string, label: string, handler: string, primary = false) => ({
  id, type: 'button', label, action: handler, primary
})

export { composeShellViewportDocument } from './catalyst-engine'

File: wasm/tests/unit/shell_viewport.rs
use super::*;

const HASH: &str = "sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";

fn model() -> Json {
    json::obj(vec![
        ("schema", json::s(MODEL_SCHEMA)),
        ("authority", json::s("none")),
        (
            "source_hashes",
            json::obj(vec![
                ("builds", json::s(HASH)),
                ("capabilities", json::s(HASH)),
                ("surfaces", json::s(HASH)),
            ]),
        ),
        (
            "shell",
            json::obj(vec![
                ("id", json::s("macos-shell")),
                ("owner", json::s("macos")),
                ("platform", json::s("macos")),
                ("manifest", json::s("macos-shell/BUILD.json")),
            ]),
        ),
        (
            "surface",
            json::obj(vec![
                ("id", json::s("desktop-16")),
                ("name", json::s("Desktop 16")),
                ("form_factor", json::s("desktop")),
                ("geometry", json::obj(vec![("width", Json::Int(1_512))])),
                ("safe_area", json::obj(vec![("top", Json::Int(0))])),
                ("corner_radii", json::obj(vec![("tl", Json::Int(8))])),
                ("chrome", Json::Arr(vec![json::s("titlebar")])),
                ("window_policy", json::s("resizable")),
            ]),
        ),
        (
            "target",
            json::obj(vec![
                ("id", json::s("aarch64-apple-darwin")),
                ("owner_ref", json::s("macos-shell/BUILD.json")),
                ("reason", json::s("structural")),
                (
                    "rungs",
                    json::obj(vec![
                        ("source", json::s("present")),
                        ("artifact", json::s("absent")),
                        ("package_install", json::s("absent")),
                        ("physical_runtime", json::s("absent")),
                    ]),
                ),
            ]),
        ),
        (
            "capability_summary",
            json::obj(vec![
                ("available", Json::Int(3)),
                ("degraded", Json::Int(1)),
                ("unavailable", Json::Int(0)),
                ("unknown", Json::Int(2)),
            ]),
        ),
        ("posture", json::s("structural-projection")),
        ("warning", json::s("STRUCTURAL PROJECTION — NOT A PHYSICAL RUN")),
        (
            "selector",
            json::obj(vec![
                ("shells", Json::Arr(vec![json::s("macos-shell"), json::s("linux-shell")])),
                ("surfaces", Json::Arr(vec![json::s("desktop-16")])),
                ("targets", Json::Arr(vec![json::s("aarch64-apple-darwin")])),
            ]),
        ),
    ])
}

fn actions(document: &Json) -> Vec<(String, bool)> {
    document
        .get("actions")
        .and_then(Json::as_array)
        .expect("actions")
        .iter()
        .map(|action| {
            (
                action.get("action").and_then(Json::as_str).unwrap_or_default().to_owned(),
                action.get("primary") == Some(&Json::Bool(true)),
            )
        })
        .collect()
}

#[test]
fn the_composed_document_is_admitted_by_the_engine() {
    let document = compose_document(&model()).expect("document");

    assert_eq!(document.get("id").and_then(Json::as_str), Some("shell-viewport"));
    assert_eq!(ugui_render::validate_document_value(&document), "[]");
}

#[test]
fn a_foreign_model_schema_refuses_rather_than_composing() {
    let Json::Obj(mut fields) = model() else { unreachable!() };

    for (name, value) in &mut fields {
        if name == "schema" {
            *value = json::s("ae-shell-viewport-model/2");
        }
    }
    assert_eq!(
        compose_document(&Json::Obj(fields)).expect_err("refusal").0,
        "shell-viewport-model-schema"
    );
}

#[test]
fn selectors_become_actions_with_the_current_choice_primary() {
    let document = compose_document(&model()).expect("document");

    assert_eq!(
        actions(&document),
        vec![
            ("shell.target.macos-shell".to_owned(), true),
            ("shell.target.linux-shell".to_owned(), false),
            ("shell.surface.desktop-16".to_owned(), true),
            ("shell.build.aarch64-apple-darwin".to_owned(), true),
            ("shell.inspect".to_owned(), false),
        ]
    );
}

#[test]
fn evidence_and_capability_counts_reach_the_status_grids() {
    let document = compose_document(&model()).expect("document");
    let sections = document.get("sections").and_then(Json::as_array).expect("sections");

    let labels = |section: &Json| {
        section
            .get("items")
            .and_then(Json::as_array)
            .expect("items")
            .iter()
            .map(|item| item.get("label").and_then(Json::as_str).unwrap_or_default().to_owned())
            .collect::<Vec<_>>()
    };

    assert_eq!(
        labels(&sections[1]),
        ["Owner", "Reason", "source", "artifact", "package_install", "physical_runtime"]
    );
    assert_eq!(labels(&sections[2]), ["available", "degraded", "unavailable", "unknown"]);
}

#[test]
fn the_receipt_carries_posture_and_source_hashes_without_claiming_authority() {
    let document = compose_document(&model()).expect("document");
    let receipt = document.get("receipt").expect("receipt");

    assert_eq!(receipt.get("schema").and_then(Json::as_str), Some(RECEIPT_SCHEMA));
    assert_eq!(receipt.get("authority").and_then(Json::as_str), Some("none"));
    assert_eq!(receipt.get("posture").and_then(Json::as_str), Some("structural-projection"));
    assert_eq!(
        receipt.get("source_hashes").and_then(|hashes| hashes.get("builds")),
        Some(&json::s(HASH))
    );
}

#[test]
fn the_native_section_carries_the_surface_geometry_the_host_observed() {
    let document = compose_document(&model()).expect("document");
    let sections = document.get("sections").and_then(Json::as_array).expect("sections");
    let native = sections[0].get("model").expect("native model");

    assert_eq!(
        sections[0].get("catalog").and_then(Json::as_str),
        Some("shell-structural-viewport")
    );
    assert_eq!(native.get("shell_id").and_then(Json::as_str), Some("macos-shell"));
    assert_eq!(native.get("window_policy").and_then(Json::as_str), Some("resizable"));
    assert_eq!(
        native.get("geometry").and_then(|geometry| geometry.get("width")),
        Some(&Json::Int(1_512))
    );
}
